### ISSUE TITLE
Intellihide: ignore maximized windows on other monitors

### ISSUE THREAD

**Issue by PeterMinin** [micheleg/dash-to-dock#285]
Currently, maximized windows cause the dock to hide even if they are on a different display that the dock. Seems like a bug.

**Comment by micheleg** [comment]
Hi,

can you please provide me with your extension an gnome shell version, as well as your settings (intellihide, which of the three options)

Thank you

**Comment by PeterMinin** [comment]
I'm sorry, after looking more carefully at the options I understood that this is an expected behavior: I have another window of the same application on the main monitor, so it fits in the description of "Only focused application's windows".
Well, then my issue is really a feature request for an "Only focused window" option, or maybe "Only focused application's windows on the same monitor" because by default secondary monitors are not considered part of the workspace, which is in fact my use case: on the second monitor I have a browser window with "background" activities like streaming music, so it's not logically connected to browser windows on the primary monitor, and thus dodging them when I activate the window on the second monitor wan't intuitive for me.

Just in case, the extension version is 51, GNOME Shell 3.14.4.

**Comment by micheleg** [comment]
I see your point. Just to be sure, when you work in the primary monitor -- i.e. the mouse focus is on a windows of another application in the primary monitor -- the dock should be shown regardless of what's going on on the secondary one. Is this the current behaviour? If not there might be a bug.

Unfortunately the perfect solution does not exist, unless one comes out with an intelligent way of defining which windows belong to the current task. The current behaviour is required  for instance if you have the main window in the primary monitor and the tolls on the secondary (e.g. in Gimp).

However, it should be trivial to change the code to filter the windows in a different way. If you want to have a look at it the logic is inside
```_intellihideFilterInteresting``` method in the ```intellihide.js``` file (https://github.com/micheleg/dash-to-dock/blob/master/intellihide.js#L277). The difficult part is only proposing meaningful options. I'll try to think of it. Whenever a tweak something I make somebody unhappy (#255).

**Comment by PeterMinin** [comment]
Yes, when an app's windows on the primary monitor don't obscure the dock, it doesn't hide even if there is a maximized window of the same app on the second monitor.

Well, I guess there is still room to add an option or two without making anyone unhappy, 3 options isn't too much (to my taste).
Thank you!

**Comment by micheleg** [comment]
The ```intellihide_options``` branch contains a tentative implementation of the suggested option. I haven't tested it because I don't have an external screen at the moment. Please have a look at it.

**Comment by PeterMinin** [comment]
My GNOME is 3.14, so I'm unable to install this version as it requires 3.18. Is it a strict requirement?
I'm on Ubuntu 15.04, so the maximum version I can get is 3.16. Maybe I'll try to upgrade to 15.10 in a couple of days.

**Comment by micheleg** [comment]
You might have checked out the wrong branch: https://github.com/micheleg/dash-to-dock/tree/intellihide_options is currently marked for 3.14 (as I'm also running 3.14 as my main environment). It will be easily ported to other versions in any case.

**Comment by PeterMinin** [comment]
Wow... Sorry, I blame my lack of sleep...
It works sometimes, but not always. I'll play around more and hopefully submit a fix and/or come up with some different options.

**Comment by PeterMinin** [comment]
Ok, it was "Always consider if belonging to the topApp" that confused me.
I'm thinking something like this would be more obvious, and flexible too:
```
v	All windows [ticks all other checkboxes]
v	Focused window
	v	Other windows of the focused application
		v	Only when the focus is in the dock's monitor
v	Top window in the dock's monitor
v	Any maximized windows
```
What do you think?

**Comment by micheleg** [comment]
So is the branch working correctly? 

For the options phrasing, I totally agree it might become confusing and I appreciate your suggestions. You seems to be adding even more options! Welcome to my nightmare of adding options without making the settings UI appears like a nuclear reactor control panel (#255).

I think the only focused window option is not very useful, and I would rather skip it. And I think the top window in the dock monitor should always be considered. Do you have a reasonable use case in which the opposite would be the desired behaviour?

I'm thinking to something like this
```
v   All windows [ticks all other checkboxes]
v   Active application's windows
    v   Only when the focus is in the dock's monitor (to be rephrased)
v   Any maximized window
```
I like the idea of the fist option selecting all the others a lot, but overall I don't like it too much because:
- No option selected would correspond to the dodge-option disabled.
- I'd like to make it clear that case 3 (Any maximized window) is a first class option by itself.

Still thinking about it...

**Comment by PeterMinin** [comment]
Yes, I just forgot to checkout it after cloning...

> No option selected would correspond to the dodge-option disabled.

You may consider linking the checkboxes to the main dodge-option toggle: when all checkboxes are unchecked, uncheck the toggle too; this would imply that the checkboxes are not disabled by the toggle, they're just cleared (remembering which were checked, though, to restore if the toggle is checked again). Or, the toggle can be removed altogether, and the "All windows" checkbox would take over its role, with a third state "some individual options checked".

> And I think the top window in the dock monitor should always be considered. Do you have a reasonable use case in which the opposite would be the desired behaviour?

Basically, when I'm focused on something in the second monitor, like chatting or browsing through music, I don't mind if the dock covers whatever I was doing on the main monitor, because I'm not doing it now. I would rather see what apps are running, "what else was I doing" other than the last active (usually maximized) window.

> I think the only focused window option is not very useful, and I would rather skip it.

Well, I try to keep all related directories/terminals/websites as tabs in one window, so usually when I'm in one window of an app, others are irrelevant, or at least not more relevant than other apps. So I think I'd be fine with "consider other windows of the active app" disabled. And here's an example of a situation when considering them is not the desired behavior. There's a maximized terminal window in the background, I'm working in some other app (let it be a non-maximized window); I then launch another instance of the terminal, non-maximized, to use a command-line calculator (I think of it as a separate application, just hosted in a terminal, thus a separate window). I'm not expecting the dock to hide because of the maximized terminal window that I have in the background, not even visible. So, I think I'm a kind of user that would benefit from this option.

I'm still thinking of a couple different ideas of the settings layout.

**Comment by micheleg** [comment]
Thanks for your additional considerations. I'll try improve the options and settings soon; as said before these options are not too difficult to implement. Regarding the single window mode, I was indeed thinking about the case of terminals, even considering the ideas of "blacklisting" them assigning a different behaviour, but this could be a bit confusing, and it might be better to just offer the simple option you are suggesting.
